backend/src/test/java/com/google/minicrm/data/AdvertiserTest.java

```diff
@@ -0,0 +1,122 @@
+import java.util.HashMap;
+import java.util.Map;
+import javax.servlet.http.HttpServletRequest;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * Provides Unit tests for the Advertiser data object.
+ */
+@RunWith(JUnit4.class)
+public final class AdvertiserTest {
+
+  private static final String TEST_USER_ID = "testUserId";
+  private static final String TEST_USER_ID_2 = "testUserId2";
+  private static final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+
+  @Before
+  public void setUp() {
+    helper.setUp();
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Test
+  public void newAdvertiser_fromValidAdvertiserAsEntity_producesEquivalentValidAdvertiser()
+      throws Exception {
+    User testUser = new User("email", "authDomain", TEST_USER_ID);
+    Advertiser advertiser = new Advertiser(testUser);
+
+    Advertiser convertedAdvertiser = new Advertiser(advertiser.asEntity());
+
+    assertEquals(advertiser, convertedAdvertiser);
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void newAdvertiser_fromInvalidEntityKind_throwsIllegalArgumentException() {
+    Entity invalidEntity = new Entity("Lead");
+
+    new Advertiser(invalidEntity);
+  }
+
+  @Test
+  public void advertiserGenerateKey_withCorrectUser_producesCorrespondingKey() {
+    User testUser = new User("email", "authDomain", TEST_USER_ID);
+    Advertiser advertiser = new Advertiser(testUser);
+
+    Key generatedKey = Advertiser.generateKey(advertiser.getUser());
+
+    assertEquals(advertiser.asEntity().getKey(), generatedKey);
+  }
+
+  @Test
+  public void advertiserGenerateKey_withDifferentUser_generatesDifferentKey()
+      throws Exception {
+    User user1 = new User("email", "authDomain", TEST_USER_ID);
+    User user2 = new User("email2", "authDomain", TEST_USER_ID_2);
+
+    Key advertiserKey1 = Advertiser.generateKey(user1);
+    Key advertiserKey2 = Advertiser.generateKey(user2);
+
+    assertNotEquals(advertiserKey1, advertiserKey2);
+  }
+
+  @Test
+  public void advertiserGenerateWebhook_withValidAdvertiser_generatesWebhookWithCorrespondingId()
+      throws Exception {
+    HttpServletRequest request = mock(HttpServletRequest.class);
+    when(request.getScheme()).thenReturn("https");
+    when(request.getServerName()).thenReturn("miniCrm.com");
+    when(request.getServerPort()).thenReturn(4200);
+    User user1 = new User("email", "authDomain", TEST_USER_ID);
+    Advertiser advertiser = new Advertiser(user1);
+
+    String webhook = advertiser.generateWebhook(request);
+
+    //get the query parameters
+    String query = webhook.substring(webhook.indexOf("?") + 1);
+    String[] keyValuePairs = query.split("&");
+    Map<String, String> params = new HashMap<String, String>();
```

Comment:
Can remove the types in HashMap to just be HashMap<>. The types are inferred from Map on the LHS.

Reply:
fixed